components/modules/pulsecnt.c

```diff
@@ -0,0 +1,732 @@
+  pulsecnt_t pc = (pulsecnt_t)lua_newuserdata(L, sizeof(pulsecnt_struct_t));
+  if (!pc) return luaL_error(L, "not enough memory");
+  luaL_getmetatable(L, "pulsecnt.pctr");
+  lua_setmetatable(L, -2);
+  pc->cb_ref = LUA_NOREF;
+  pc->self_ref = LUA_NOREF;
+  pc->is_initted = false;
+  pc->is_debug = false;
+  pc->counter = 99;
+  pc->unit = unit; // default to 0
+
+  //get the lua function reference
+  if (isCallback) {
+    luaL_unref(L, LUA_REGISTRYINDEX, pc->cb_ref);
+    lua_pushvalue(L, stack);
+    pc->cb_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  }
+
+  // store in our global static pulsecnt_selfs array for later reference during callback
+  // where we only know the unit number 
+  pulsecnt_selfs[unit] = pc;
+
+  // Get is_debug -- 3rd arg optional
+  ++stack;
+  bool is_debug = luaL_optbool(L, stack, false);
+  if (is_debug == 1) {
+    pc->is_debug = true;
+  }
+
+  if (pc->is_debug) ESP_LOGI("pulsecnt", "Created obj for unit %d with callback ref of %d", pc->unit, pc->cb_ref );
+
+  return 1;
+}
+
+// Lua: pulsecnt:testCb( )
+// This tests the callback where you can mimic in your code a sample callback as 
+// part of your debugging process. Make sure to set your callback during the create() call.
+static int pulsecnt_testCb(lua_State* L)
+{
+  // return 0;
+
+  pulsecnt_t pc = pulsecnt_get(L, 1);
+
+  // get a self reference if we don't have one
+  if (pc->self_ref == LUA_NOREF) {
+    lua_pushvalue(L, 1);
+    pc->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  }
+
+  if (pc->cb_ref == LUA_NOREF) {
+    luaL_error(L, "Callback not registered");
+  }
+
+  // post using lua task posting technique
+  // on lua_open we set pulsecnt_task_id as a method which gets called
+  // by Lua after task_post_high with reference to this self object and then we can steal the 
+  // callback_ref and then it gets called by lua_call where we get to add our args
+  // task_post_low(pulsecnt_task_id, (task_param_t)pc);
+  task_post_low(pulsecnt_task_id, ( (0xffu) << 8) | (pc->unit) );
+  
+
+  return 0;
+}
+
+// Lua: pulsecnt:getCnt( )
+// Get's the pulse counter for a unit.
+static int pulsecnt_getCnt(lua_State* L)
+{
+  pulsecnt_t pc = pulsecnt_get(L, 1);
+
+  int16_t count = 0;
+  pcnt_get_counter_value(pc->unit, &count);
+  if (pc->is_debug) ESP_LOGI("pulsecnt", "Got ctr val for unit %d with count of %d", pc->unit, count );
+  pc->counter = count;
+
+  lua_pushinteger(L, pc->counter);
+  return 1;
+}
+
+// Lua: pulsecnt:clear( )
+// Clear the pulse counter, i.e. set it back to 0.
+static int pulsecnt_clear(lua_State* L)
+{
+  pulsecnt_t pc = pulsecnt_get(L, 1);
+  pcnt_counter_clear(pc->unit);
+  int16_t count = 0;
+  pcnt_get_counter_value(pc->unit, &count);
+  if (pc->is_debug) ESP_LOGI("pulsecnt", "Cleared ctr for unit %d with new count of %d", pc->unit, count );
+  pc->counter = count;
+  lua_pushinteger(L, pc->counter);
+  return 1;
+}
+
+// Lua: pulsecnt:unregister( self )
+static int pulsecnt_unregister(lua_State* L){
```

Comment:
Will probably never be called from the garbage collector when the object is registered in `self_ref`.

Reply:
I'm following the design pattern in the tmr module. It too has callbacks that are modeled similar to my callbacks and establish the self_ref and clean it up in the unregister method. Although, now that you've pointed this out, I think the reason tmr has to keep track is for one time timers where your tmr object could go out of scope, you get your callback, and then tmr cleans up the ref. I don't really need that, so I could eliminate self_ref completely.